Change schedulers to chainable form

vincentqb opened the pull request:
Enable chainable schedulers as requested in #13022 by implementing the changes mentioned below from [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513370208).

* Changing the behavior of schedulers to the chainable formula when available
* Using the closed form whenever epoch is different from None until the next release with a deprecation warning
* Making `get_computed_values` the supported way of obtaining the last computed learning rate by the scheduler (see [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513940729) for new syntax)
* Returning a deprecation warning when invoking the undocumented get_lr function (see [comment](https://github.com/pytorch/pytorch/pull/21800#discussion_r294305485)) referring to `get_computed_values`, and deprecating it in the next release.
* `CosineAnnealingWarmRestart` still takes an epoch parameter as it is the only one with a mechanic relying on fractional epoch
* `MultiplicativeLR` is consumes a function providing the multiplicative factor at each epoch. It mimics `LambdaLR` in its syntax. 

# #20527

### Before

The user calls scheduler with a constant epoch either across loops or in the same loop.
```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)

# Scheduler with sometimes-constant epoch number
for epoch in [0, 0, 1, 1, 2, 2, 3, 3]:
  lr_scheduler.step(epoch)
  print(optimizer.param_groups[0]['lr'])
```

### After

If the user wants to step
```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)

last_epoch = -1
for epoch in [0, 0, 1, 1, 2, 2, 3, 3]:

  # Check if epoch number has changed manually
  if epoch-last_epoch > 0:
    lr_scheduler.step()
  last_epoch = epoch

  print(epoch, scheduler.get_computed_values())
```

# #22107

### Before

```
import torch
from torchvision.models import resnet18
net = resnet18()

optimizer = torch.optim.SGD(net.parameters(), 0.1)
scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)

for i in range(10):
  # Scheduler computes and returns new learning rate, leading to unexpected behavior
  print(i, scheduler.get_lr())
  scheduler.step()
```

### After

```
import torch
from torchvision.models import resnet18

net = resnet18()
optimizer = torch.optim.SGD(net.parameters(), 0.1)
lr_scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)

for i in range(10):
    # Returns last computed learning rate by scheduler
    print(i, lr_scheduler.get_computed_values())
    lr_scheduler.step()
```

Differential Revision: [D17349760](https://our.internmc.facebook.com/intern/diff/D17349760)

vincentqb committed:
Enable chainable schedulers as requested in #13022 by implementing the changes mentioned below from [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513370208).

* Changing the behavior of schedulers to the chainable formula when available
* Using the closed form whenever epoch is different from None until the next release with a deprecation warning
* Making get_computed_values the supported way of obtaining the last computed learning rate by the scheduler (see [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513940729) for new syntax)
* Returning a deprecation warning when invoking the undocumented get_lr function (see [comment](https://github.com/pytorch/pytorch/pull/21800#discussion_r294305485)) referring to get_computed_values, and deprecating it in the next release.
* `CosineAnnealingWarmRestart` still takes an epoch parameter as it is the only one with a mechanic relying on fractional epoch

vincentqb committed:
Update on "Change schedulers to chainable form"

Enable chainable schedulers as requested in #13022 by implementing the changes mentioned below from [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513370208).

Stack from [ghstack](https://github.com/ezyang/ghstack):

* Changing the behavior of schedulers to the chainable formula when available
* Using the closed form whenever epoch is different from None until the next release with a deprecation warning
* Making get_computed_values the supported way of obtaining the last computed learning rate by the scheduler (see [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513940729) for new syntax)
* Returning a deprecation warning when invoking the undocumented get_lr function (see [comment](https://github.com/pytorch/pytorch/pull/21800#discussion_r294305485)) referring to get_computed_values, and deprecating it in the next release.
* `CosineAnnealingWarmRestart` still takes an epoch parameter as it is the only one with a mechanic relying on fractional epoch

soumith commented on torch/optim/lr_scheduler.py:
```diff
@@ -1,23 +1,42 @@
 
 
+EPOCH_DEPRECATION_WARNING = (
+    "The behavior of schedulers has been changed from their closed form to their "
```
this statement is not comprehensible (I have no idea what it means after reading it)

vincentqb replied:
Thanks for the feedback. Please let me know if the new version is clearer.

vincentqb replied:
Thanks for the feedback. Please let me know if the new version is clearer.

soumith commented on torch/optim/lr_scheduler.py:
```diff
@@ -59,9 +78,34 @@ def load_state_dict(self, state_dict):
+        try:
+            return self._last_computed_values
+        except AttributeError:
+            raise RuntimeError("Scheduler needs to step to get last computed values.")
```
You haven't called `.step()` even once on the Scheduler, so there are no last computed values

vincentqb replied:
Indeed, I'm capturing the `AttributeError` and returning a `RuntimeError` telling the user to take a step with scheduler. I've clarified the comment. Is that what you meant?

vincentqb replied:
`step` is called in `__init__`.

vincentqb committed:
Update on "Change schedulers to chainable form"

Enable chainable schedulers as requested in #13022 by implementing the changes mentioned below from [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513370208).

Stack from [ghstack](https://github.com/ezyang/ghstack):

* Changing the behavior of schedulers to the chainable formula when available
* Using the closed form whenever epoch is different from None until the next release with a deprecation warning
* Making get_computed_values the supported way of obtaining the last computed learning rate by the scheduler (see [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513940729) for new syntax)
* Returning a deprecation warning when invoking the undocumented get_lr function (see [comment](https://github.com/pytorch/pytorch/pull/21800#discussion_r294305485)) referring to get_computed_values, and deprecating it in the next release.
* `CosineAnnealingWarmRestart` still takes an epoch parameter as it is the only one with a mechanic relying on fractional epoch
* `LambdaLR` maintains its non-chainable form in this PR

ezyang commented:
Lint failures are real

ezyang commented:
First, many apologies for the delayed review. Fell off the radar between release and then travel.

I'm reading the PR description and thinking to myself that, if I were an end user trying to understand what happened to LR schedulers after this diff, this description wouldn't be enough. One of the things that is really useful in the release notes is "before and after" code samples, and I think those would also be very helpful here (and help front-loading the work for release notes the next time we release). https://github.com/pytorch/pytorch/releases

fmassa commented on torch/optim/lr_scheduler.py:
```diff
@@ -82,12 +121,23 @@ def step(self, epoch=None):
+            self.last_epoch = self.last_epoch + 1
+            values = self._compute_values()
+        else:
+            warnings.warn(EPOCH_DEPRECATION_WARNING, DeprecationWarning)
```
is `DeprecationWarning` ignored by default? https://github.com/pytorch/pytorch/blob/92750acb88ec5539658b1c51ee27a648ce92a33a/torch/nn/modules/container.py#L15-L17

vincentqb replied:
I see the deprecation warning being raised when running the `test/test_optim.py` without warning filters modified. Do you know the context of that warning?

vincentqb commented on torch/optim/lr_scheduler.py:
```diff
@@ -26,6 +43,7 @@ def __init__(self, optimizer, last_epoch=-1):
         self.base_lrs = list(map(lambda group: group['initial_lr'], optimizer.param_groups))
         self.last_epoch = last_epoch
```
I'm assuming the `-1` value is a leftover from when shedulers were called at the beginning of their epoch, and so the previous epoch of 0 would have been -1. Since we are now stepping schedulers at the end of the iteration, it is indeed confusing. It would make sense to default to `last_epoch=0` instead.

vincentqb replied:
> I think there should be a clear description of what `last_epoch == -1` means, it would just be nice if it were spelled out explicitly.

`last_epoch` is a scheduler-specific concept, and many schedulers in chainable form do not need an explicit `last_epoch`. The concept is supposed to say "This is the epoch that was last invoked. Before the scheduler's creation, `last_epoch` is -1, immediately after creation `last_epoch` is then 0, etc.

In fact, we should also deprecate `last_epoch` from `_LRScheduler`'s initialization. This means that other schedulers will need to re-implement that logic per-scheduler.

Thoughts?

> This is especially because previously we set `last_epoch == -1` to `0`, and now we don't anymore.

I consider that setting `last_epoch` to 0 before was a bug, see [here](https://github.com/pytorch/pytorch/pull/23306#discussion_r307434811) for our previous discussion on the meaning of the change.

vincentqb replied:
For reference: the impact of `last_epoch=-1` is discussed in #7889.

vincentqb commented on torch/optim/lr_scheduler.py:
```diff
@@ -59,9 +76,31 @@ def load_state_dict(self, state_dict):
 
-    def get_lr(self):
+    def get_computed_values(self):
+        # Return last computed learning rate by current scheduler
```
Thanks for pointing this out!

vincentqb commented on torch/optim/lr_scheduler.py:
```diff
@@ -59,9 +76,31 @@ def load_state_dict(self, state_dict):
+                "`get_computed_values()`. Prior to this, `get_lr()` was an"
+                "internal method to compute new scheduler values within"
+                "`step()`, leading to unexpected behavior when invoked"
+                "directly.", DeprecationWarning)
```
`get_lr` was not a reliable way of getting the current learning rate. The code fixes that issue by simply returning the last computed value by the scheduler.

In order to make this transparent, we've decide to replace the name `get_lr` which could imply "get current learning rate" to `get_computed_values` to make it clear this returns the last computed values by the scheduler.

I've simplified the message: as far as the user is concerned, the method is simply renamed.

vincentqb commented on torch/optim/lr_scheduler.py:
```diff
@@ -82,12 +121,23 @@ def step(self, epoch=None):
-            epoch = self.last_epoch + 1
-        self.last_epoch = epoch
-        for param_group, lr in zip(self.optimizer.param_groups, self.get_lr()):
+            self.last_epoch = self.last_epoch + 1
```
thanks for pointing this out!

vincentqb commented on torch/optim/lr_scheduler.py:
```diff
@@ -82,12 +121,23 @@ def step(self, epoch=None):
+            try:
+                values = self._compute_values_closed_form()
+            except NotImplementedError:
+                values = self._compute_values()
```
We are not ignoring the epoch, see the line before `try`: `self.last_epoch = epoch`.

We could show the warning only when actually using the closed form, since some schedulers didn't have a closed form in the first place. Thoughts?

vincentqb commented on torch/optim/lr_scheduler.py:
```diff
@@ -59,9 +76,31 @@ def load_state_dict(self, state_dict):
+        try:
+            return self._last_computed_values
+        except AttributeError:
+            raise RuntimeError("Run `step()` at least once before `get_computed_values()`.")
```
Nope, will remove :)

vincentqb replied:
Indeed :) Removing try clause.

vincentqb commented on torch/optim/lr_scheduler.py:
```diff
@@ -705,10 +794,21 @@ def __init__(self, optimizer, T_0, T_mult=1, eta_min=0, last_epoch=-1):
-        super(CosineAnnealingWarmRestarts, self).__init__(optimizer, last_epoch)
+
+        # TODO Invoke super instead of repeating the parent __init__ with epoch passed to self.step
+        # super(CosineAnnealingWarmRestarts, self).__init__(optimizer, last_epoch)
```
Agree. Fix upcoming.

vincentqb committed:
Update on "Change schedulers to chainable form"

Enable chainable schedulers as requested in #13022 by implementing the changes mentioned below from [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513370208).

* Changing the behavior of schedulers to the chainable formula when available
* Using the closed form whenever epoch is different from None until the next release with a deprecation warning
* Making get_computed_values the supported way of obtaining the last computed learning rate by the scheduler (see [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513940729) for new syntax)
* Returning a deprecation warning when invoking the undocumented get_lr function (see [comment](https://github.com/pytorch/pytorch/pull/21800#discussion_r294305485)) referring to get_computed_values, and deprecating it in the next release.
* `CosineAnnealingWarmRestart` still takes an epoch parameter as it is the only one with a mechanic relying on fractional epoch
* `LambdaLR` maintains its non-chainable form in this PR

# #20527

### Before
```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)

for _ in range(10):
  # Scheduler computes rate for epoch 2 using closed form formula
  lr_scheduler.step(2)
  print(optimizer.param_groups[0]['lr'])
```

### After

If the user wants to step
```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)

for _ in range(10):
    # Scheduler computes new iteration of learning rate, without epoch parameter
    lr_scheduler.step()
    print(optimizer.param_groups[0]['lr'])
```

or if the user wants an explicit constant function and use the scheduler framework
```
lr_scheduler = LambdaLR(optimizer, lr_lambda=lambda epoch: 0.5)

for _ in range(10):
    # The new constant scheduler simply returns constant learning rate
    lr_scheduler.step()
    print(lr_scheduler.get_computed_values())
```

# #22107

### Before

```
import torch
from torchvision.models import resnet18
net = resnet18()

optimizer = torch.optim.SGD(net.parameters(), 0.1)
scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)

for i in range(10):
  # Scheduler computes and returns new learning rate, leading to unexpected behavior
  print(i, scheduler.get_lr())
  scheduler.step()
```

### After

```
import torch
from torchvision.models import resnet18

net = resnet18()
optimizer = torch.optim.SGD(net.parameters(), 0.1)
lr_scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)

for i in range(10):
    # Returns last computed learning rate by scheduler
    print(i, lr_scheduler.get_computed_values())
    lr_scheduler.step()
```

vincentqb committed:
Update on "Change schedulers to chainable form"

Enable chainable schedulers as requested in #13022 by implementing the changes mentioned below from [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513370208).

* Changing the behavior of schedulers to the chainable formula when available
* Using the closed form whenever epoch is different from None until the next release with a deprecation warning
* Making get_computed_values the supported way of obtaining the last computed learning rate by the scheduler (see [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513940729) for new syntax)
* Returning a deprecation warning when invoking the undocumented get_lr function (see [comment](https://github.com/pytorch/pytorch/pull/21800#discussion_r294305485)) referring to get_computed_values, and deprecating it in the next release.
* `CosineAnnealingWarmRestart` still takes an epoch parameter as it is the only one with a mechanic relying on fractional epoch
* `LambdaLR` maintains its non-chainable form in this PR

# #20527

### Before
```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)

for _ in range(10):
  # Scheduler computes rate for epoch 2 using closed form formula
  lr_scheduler.step(2)
  print(optimizer.param_groups[0]['lr'])
```

### After

If the user wants to step
```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)

for _ in range(10):
    # Scheduler computes new iteration of learning rate, without epoch parameter
    lr_scheduler.step()
    print(optimizer.param_groups[0]['lr'])
```

or if the user wants an explicit constant function and use the scheduler framework
```
lr_scheduler = LambdaLR(optimizer, lr_lambda=lambda epoch: 0.5)

for _ in range(10):
    # The new constant scheduler simply returns constant learning rate
    lr_scheduler.step()
    print(lr_scheduler.get_computed_values())
```

# #22107

### Before

```
import torch
from torchvision.models import resnet18
net = resnet18()

optimizer = torch.optim.SGD(net.parameters(), 0.1)
scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)

for i in range(10):
  # Scheduler computes and returns new learning rate, leading to unexpected behavior
  print(i, scheduler.get_lr())
  scheduler.step()
```

### After

```
import torch
from torchvision.models import resnet18

net = resnet18()
optimizer = torch.optim.SGD(net.parameters(), 0.1)
lr_scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)

for i in range(10):
    # Returns last computed learning rate by scheduler
    print(i, lr_scheduler.get_computed_values())
    lr_scheduler.step()
```

ezyang commented:
Re the first Before/After, are you sure the first After actually does the same thing as the code before? Usually users want to know how to exactly replicate the behavior they had before; if they were "Doing it Wrong(TM)" we also need to explain how they were doing it wrong.

ezyang commented:
OK, on further reflection, I think the Before/After for #20527 is missing the point of the issue. See in the original issue:

> I just expect the schedule to be consistent with the epoch, as if there's a time table specifying the learning rate at each epoch. Otherwise the epoch parameter is misleading.
>
> To answer your question; yes I don't want it to reduce or increase if it is called with the same epoch value.

@jasam-sheja you would be able to say better, but I think what your actual code looked like was something like:

```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)

for epoch in range(10):
  # ... some large amount of code ...
  lr_scheduler.step(epoch)
  # ... some large amount of code ...
  lr_scheduler.step(epoch)
  # ... some large amount of code ...
```

Before, `step()` with an explicit epoch was idempotent. People shouldn't really do it, but it would get harmless discarded if you did. I'm guessing now it's not? So you need to now remove the redundant steps, I guess. I feel like an old version of this patch warned when this happen, no?

vincentqb commented:
> Before, `step()` with an explicit epoch was idempotent. People shouldn't really do it, but it would get harmless discarded if you did. I'm guessing now it's not? So you need to now remove the redundant steps, I guess. I feel like an old version of this patch warned when this happen, no?

I'm not sure if people were doing it, but one use case I could think of is: a training loop is at `iter` iterations, which corresponds to `epoch = iter//length_data`, and then the user can call `scheduler.step(epoch)` at each iteration instead of only at each epoch.

vincentqb commented:
> Before, `step()` with an explicit epoch was idempotent. People shouldn't really do it, but it would get harmless discarded if you did. I'm guessing now it's not? So you need to now remove the redundant steps, I guess. 

If the epoch is specified and a closed form exists, the closed form will be used, and so the next call will be discarded with a deprecation warning.

Otherwise, the next calls with the same epoch won't be discarded, but a deprecation warning will still be raised. 

> I feel like an old version of this patch warned when this happen, no?

Whenever a user calls `step` with `epoch` a deprecation error will be raised. Is this what you mean?

vincentqb committed:
Update on "Change schedulers to chainable form"

Enable chainable schedulers as requested in #13022 by implementing the changes mentioned below from [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513370208).

* Changing the behavior of schedulers to the chainable formula when available
* Using the closed form whenever epoch is different from None until the next release with a deprecation warning
* Making get_computed_values the supported way of obtaining the last computed learning rate by the scheduler (see [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513940729) for new syntax)
* Returning a deprecation warning when invoking the undocumented get_lr function (see [comment](https://github.com/pytorch/pytorch/pull/21800#discussion_r294305485)) referring to get_computed_values, and deprecating it in the next release.
* `CosineAnnealingWarmRestart` still takes an epoch parameter as it is the only one with a mechanic relying on fractional epoch
* `LambdaLR` maintains its non-chainable form in this PR

# #20527

### Before

The user calls scheduler with a constant epoch either across loops or in the same loop.
```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)

for _ in range(10):
  # Scheduler computes rate for epoch 2 using closed form formula
  lr_scheduler.step(2)
  lr_scheduler.step(2)
  print(optimizer.param_groups[0]['lr'])
```

### After

If the user wants to step
```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)

for _ in range(10):
    # Scheduler computes new iteration of learning rate, without epoch parameter
    lr_scheduler.step()
    print(optimizer.param_groups[0]['lr'])
```

or if the user wants an explicit constant function and use the scheduler framework
```
lr_scheduler = LambdaLR(optimizer, lr_lambda=lambda epoch: 0.5)

for _ in range(10):
    # The new constant scheduler simply returns constant learning rate
    lr_scheduler.step()
    print(lr_scheduler.get_computed_values())
```

# #22107

### Before

```
import torch
from torchvision.models import resnet18
net = resnet18()

optimizer = torch.optim.SGD(net.parameters(), 0.1)
scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)

for i in range(10):
  # Scheduler computes and returns new learning rate, leading to unexpected behavior
  print(i, scheduler.get_lr())
  scheduler.step()
```

### After

```
import torch
from torchvision.models import resnet18

net = resnet18()
optimizer = torch.optim.SGD(net.parameters(), 0.1)
lr_scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)

for i in range(10):
    # Returns last computed learning rate by scheduler
    print(i, lr_scheduler.get_computed_values())
    lr_scheduler.step()
```

vincentqb commented:
> This was intentional in the sense that I wanted to discuss with @fmassa before adding the API had in mind.

Chainable LambdaLR updated.

vincentqb committed:
Update on "Change schedulers to chainable form"

Enable chainable schedulers as requested in #13022 by implementing the changes mentioned below from [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513370208).

* Changing the behavior of schedulers to the chainable formula when available
* Using the closed form whenever epoch is different from None until the next release with a deprecation warning
* Making `get_computed_values` the supported way of obtaining the last computed learning rate by the scheduler (see [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513940729) for new syntax)
* Returning a deprecation warning when invoking the undocumented get_lr function (see [comment](https://github.com/pytorch/pytorch/pull/21800#discussion_r294305485)) referring to `get_computed_values`, and deprecating it in the next release.
* `CosineAnnealingWarmRestart` still takes an epoch parameter as it is the only one with a mechanic relying on fractional epoch
* `LambdaLR` now has a toggle to switch between chainable and non-chainable form. It currently defaults to non-chainable, but also returns a FutureWarning to default to chainable form later like most other schedulers.

# #20527

### Before

The user calls scheduler with a constant epoch either across loops or in the same loop.
```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)

for _ in range(10):
  # Scheduler computes rate for epoch 2 using closed form formula
  lr_scheduler.step(2)
  lr_scheduler.step(2)
  print(optimizer.param_groups[0]['lr'])
```

### After

If the user wants to step
```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)

for _ in range(10):
    # Scheduler computes new iteration of learning rate, without epoch parameter
    lr_scheduler.step()
    print(optimizer.param_groups[0]['lr'])
```

or if the user wants an explicit constant function and use the scheduler framework
```
lr_scheduler = LambdaLR(optimizer, lr_lambda=lambda epoch: 0.5)

for _ in range(10):
    # The new constant scheduler simply returns constant learning rate
    lr_scheduler.step()
    print(lr_scheduler.get_computed_values())
```

# #22107

### Before

```
import torch
from torchvision.models import resnet18
net = resnet18()

optimizer = torch.optim.SGD(net.parameters(), 0.1)
scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)

for i in range(10):
  # Scheduler computes and returns new learning rate, leading to unexpected behavior
  print(i, scheduler.get_lr())
  scheduler.step()
```

### After

```
import torch
from torchvision.models import resnet18

net = resnet18()
optimizer = torch.optim.SGD(net.parameters(), 0.1)
lr_scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)

for i in range(10):
    # Returns last computed learning rate by scheduler
    print(i, lr_scheduler.get_computed_values())
    lr_scheduler.step()
```

vincentqb committed:
Update on "Change schedulers to chainable form"

Enable chainable schedulers as requested in #13022 by implementing the changes mentioned below from [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513370208).

* Changing the behavior of schedulers to the chainable formula when available
* Using the closed form whenever epoch is different from None until the next release with a deprecation warning
* Making `get_computed_values` the supported way of obtaining the last computed learning rate by the scheduler (see [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513940729) for new syntax)
* Returning a deprecation warning when invoking the undocumented get_lr function (see [comment](https://github.com/pytorch/pytorch/pull/21800#discussion_r294305485)) referring to `get_computed_values`, and deprecating it in the next release.
* `CosineAnnealingWarmRestart` still takes an epoch parameter as it is the only one with a mechanic relying on fractional epoch
* `LambdaLR` now has a toggle to switch between chainable and non-chainable form. It currently defaults to non-chainable, but also returns a FutureWarning to default to chainable form later like most other schedulers.

# #20527

### Before

The user calls scheduler with a constant epoch either across loops or in the same loop.
```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)

# Scheduler with sometimes-constant epoch number
for epoch in [0, 0, 1, 1, 2, 2, 3, 3]:
  lr_scheduler.step(epoch)
  print(optimizer.param_groups[0]['lr'])
```

### After

If the user wants to step
```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)

last_epoch = -1
for epoch in [0, 0, 1, 1, 2, 2, 3, 3]:
  if epoch-last_epoch > 0:
    lr_scheduler.step()
  last_epoch = epoch
  print(epoch, scheduler.get_computed_values())
```

# #22107

### Before

```
import torch
from torchvision.models import resnet18
net = resnet18()

optimizer = torch.optim.SGD(net.parameters(), 0.1)
scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)

for i in range(10):
  # Scheduler computes and returns new learning rate, leading to unexpected behavior
  print(i, scheduler.get_lr())
  scheduler.step()
```

### After

```
import torch
from torchvision.models import resnet18

net = resnet18()
optimizer = torch.optim.SGD(net.parameters(), 0.1)
lr_scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)

for i in range(10):
    # Returns last computed learning rate by scheduler
    print(i, lr_scheduler.get_computed_values())
    lr_scheduler.step()
```

vincentqb commented:
> > This was intentional in the sense that I wanted to discuss with @fmassa before adding the API had in mind.
> 
> Chainable LambdaLR updated.

Based on feedback from @ezyang, I've left LambdaLR untouched, and introduced `MultiplicativeLR` for the chainable form of LambdaLR.

ezyang reviewed (approved):
I'm going to approve this in the interest of moving things along.

vincentqb committed:
Update on "Change schedulers to chainable form"

Enable chainable schedulers as requested in #13022 by implementing the changes mentioned below from [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513370208).

* Changing the behavior of schedulers to the chainable formula when available
* Using the closed form whenever epoch is different from None until the next release with a deprecation warning
* Making `get_computed_values` the supported way of obtaining the last computed learning rate by the scheduler (see [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513940729) for new syntax)
* Returning a deprecation warning when invoking the undocumented get_lr function (see [comment](https://github.com/pytorch/pytorch/pull/21800#discussion_r294305485)) referring to `get_computed_values`, and deprecating it in the next release.
* `CosineAnnealingWarmRestart` still takes an epoch parameter as it is the only one with a mechanic relying on fractional epoch
* `MultiplicativeLR` is consumes a function providing the multiplicative factor at each epoch. It mimics `LambdaLR` in its syntax. 

# #20527

### Before

The user calls scheduler with a constant epoch either across loops or in the same loop.
```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)

# Scheduler with sometimes-constant epoch number
for epoch in [0, 0, 1, 1, 2, 2, 3, 3]:
  lr_scheduler.step(epoch)
  print(optimizer.param_groups[0]['lr'])
```

### After

If the user wants to step
```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)

last_epoch = -1
for epoch in [0, 0, 1, 1, 2, 2, 3, 3]:

  # Check if epoch number has changed manually
  if epoch-last_epoch > 0:
    lr_scheduler.step()
  last_epoch = epoch

  print(epoch, scheduler.get_computed_values())
```

# #22107

### Before

```
import torch
from torchvision.models import resnet18
net = resnet18()

optimizer = torch.optim.SGD(net.parameters(), 0.1)
scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)

for i in range(10):
  # Scheduler computes and returns new learning rate, leading to unexpected behavior
  print(i, scheduler.get_lr())
  scheduler.step()
```

### After

```
import torch
from torchvision.models import resnet18

net = resnet18()
optimizer = torch.optim.SGD(net.parameters(), 0.1)
lr_scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)

for i in range(10):
    # Returns last computed learning rate by scheduler
    print(i, lr_scheduler.get_computed_values())
    lr_scheduler.step()
```

Differential Revision: [D17349760](https://our.internmc.facebook.com/intern/diff/D17349760)

vincentqb commented:
This was reverted and replaced by #26423
